torch/csrc/PyInterpreter.cpp

```diff
@@ -18,17 +18,20 @@ namespace {
 // because passing in constexpr char* as template argument breaks some
 // versions of MSVC that are being used internally at Meta.
 // MSVC 14.16.27023 (vs2017_15.9)
-#define CONCRETE_TRACE_CUDA(func_name, ...)                           \
-  at::impl::MaybeSetTLSOnEntryGuard guard;                            \
-  if (Py_IsInitialized()) {                                           \
-    pybind11::gil_scoped_acquire gil;                                 \
-    try {                                                             \
-      py::module mod = py::module::import("torch.utils._cuda_trace"); \
-      py::object hook = mod.attr(func_name).attr("fire_callbacks");   \
-      hook(__VA_ARGS__);                                              \
-    } catch (const std::exception& e) {                               \
-      LOG(ERROR) << "CUDA trace hook execution failed: " << e.what(); \
-    }                                                                 \
+#define CONCRETE_GPU_TRACE(device_type, func_name, ...)                     \
+  at::impl::MaybeSetTLSOnEntryGuard guard;                                  \
+  if (Py_IsInitialized()) {                                                 \
+    pybind11::gil_scoped_acquire gil;                                       \
+    try {                                                                   \
+      std::string module_name =                                             \
+          "torch." + at::DeviceTypeName(device_type, true) + "._gpu_trace"; \
```

Comment:
An assumption: the callback functions will be located in `torch.xxx._gpu_trace.py` for each backend.

Reply:
You can use https://github.com/pytorch/pytorch/blob/be0bdf111ccfb403637e40dc26199c116b18cf73/torch/_utils.py#L891 here to simplify the logic

Reply:
Do you mean like this,
```cpp
try {                                                                    \
  py::module utils_mod = py::module::import("torch._utils");             \
  py::object get_device_module = utils_mod.attr("_get_device_module");   \
  py::object hook = get_device_module(DeviceTypeName(device_type, true)) \
                          .attr("_gpu_trace")                            \
                          .attr(func_name)                               \
                          .attr("fire_callbacks");                       \
    hook(__VA_ARGS__);                                                   \
  } catch (const std::exception& e) {                                    \
    LOG(ERROR) << device_type                                            \
               << " trace hook execution failed: " << e.what();          \
  }                                                                      \
```

Reply:
Option 2:
```cpp
try {                                                                       \
    std::string module_name = "torch." + DeviceTypeName(device_type, true); \
    py::module mod = py::module::import(module_name.c_str());               \
    py::object hook =                                                       \
        mod.attr("_gpu_trace").attr(func_name).attr("fire_callbacks");      \
    hook(__VA_ARGS__);                                                      \
  } catch (const std::exception& e) {                                       \
    LOG(ERROR) << device_type                                               \
               << " trace hook execution failed: " << e.what();             \
  }                                                                         \
```
option 2 is more concise, and Pybind can also raise a friendly error.
Which one do you prefer?